### PR TITLE
Add `HEAD` & `TRACE` Support

### PR DESCRIPTION
**Pull request by ItzNotABug**
The runtimes use an external library to manage the http server where not all the underlying libraries/modules support `TRACE` method. Below is a list of the one's I've tested with findings -

---
### Tested
* Dart-3.3
* Bun-1.0
* Ruby-3.3
* Kotlin 1.9
* Deno-1.40
* Node-21.0
* Python-3.12
* Python-ML-3.11
* Swift-5.9, 5.8, 5.5

---
### `TRACE` not supported -
* PHP
* C++ - 17, 20 (Drogon Library)
   There might be a possibility of using a Middleware approach but I am not entirely sure on that

### CODE REVIEW

File: tests/Base.php [hunk 1]
```diff
@@ -213,6 +213,15 @@ public function testRequestMethod(): void
         $response = $this->execute(method: 'PATCH', headers: ['x-action' => 'requestMethod']);
         self::assertEquals(200, $response['code']);
         self::assertEquals('PATCH', $response['body']);
+
+        $response = $this->execute(method: 'HEAD', headers: ['x-action' => 'requestMethod']);
+        self::assertEquals(200, $response['code']);
+        // Some runtimes either return '', while some return 'HEAD'
+        // self::assertEquals('', $response['body']);
```
Comment (Meldiron) [inline tests/Base.php:220]: Let's enable this test and allow both. Refference to allowing 2 options in a test: https://github.com/ItzNotABug/open-runtimes/blob/684549e12f49def843cc2162428ab2f7c5588f33/tests/Base.php#L241

File: tests/Base.php [hunk 2]
```diff
@@ -213,6 +213,15 @@ public function testRequestMethod(): void
         $response = $this->execute(method: 'PATCH', headers: ['x-action' => 'requestMethod']);
```
Comment (Meldiron) [inline tests/Base.php:213]: Above this test is `OPTIONS`, seems to only be problematic on C++. Let's please enable it as part of this PR. Ill also take a look at all failing tests, and give helping hand. If neither of us can figure it out, we can again disable the test, but specifically for C++ only.

File: tests/Base.php (hunk 1 above)
Comment (ItzNotABug) [reply tests/Base.php:220]: Makes sense. I'll enable this one.

File: tests/Base.php (hunk 2 above)
Comment (ItzNotABug) [reply tests/Base.php:213]: Sure thing.

**Comment by ItzNotABug** [comment]
I guess the Kotlin tests failed too, I'll double check the `javalin`'s impl.

File: tests/Base.php [hunk 3]
```diff
@@ -209,10 +209,20 @@ public function testRequestMethod(): void
...
+        self::assertContains($response['body'], ['OPTIONS', '']);
 
         $response = $this->execute(method: 'PATCH', headers: ['x-action' => 'requestMethod']);
         self::assertEquals(200, $response['code']);
         self::assertEquals('PATCH', $response['body']);
+
+        $response = $this->execute(method: 'HEAD', headers: ['x-action' => 'requestMethod']);
+        self::assertEquals(200, $response['code']);
+        //  var_dump says the body is a boolean in C++.
+        self::assertContains($response['body'], ['HEAD', '', false]);
```
Comment (ItzNotABug) [inline tests/Base.php:221]: @Meldiron,
One thing to note here is that if I do below, the body is properly an empty string -
```php
self::assertContains(trim($response['body']), ['HEAD', '']);
```
else its a boolean, atleast as per the `var_dump`.

**Comment by byawitz** [comment]
@ItzNotABug the Kotlin would is a x64 image, to run it on an arm machines as Apple `mX` processors run it with the `--platform linux/amd64` flag 

image

**Comment by ItzNotABug** [comment]
@byawitz Thanks for the tip!
It worked & I was able to run the tests!

File: tests/Base.php (hunk 3 above)
Comment (ItzNotABug) [reply tests/Base.php:221]: Something similar happens on the Kotlin side as well -
```php
["body"]=> bool(false)
```

File: tests/Base.php [hunk 4]
```diff
@@ -10,8 +10,11 @@
 
 class Base extends TestCase
 {
+    private string $runtime = '';
+
     public function setUp(): void
     {
+        $this->runtime = \explode('-', \getenv('RUNTIME'), 2)[0] ?? '';
```
Comment (Meldiron) [inline tests/Base.php:17]: This wont work well for `python-ml-3.11`, for example. We need to split by `-`, remove last from array, and join back with `-`. With that logic we get `python-ml` proplerly.

File: tests/Base.php [hunk 5]
```diff
@@ -207,14 +212,40 @@ public function testRequestMethod(): void
...
+
+        $response = $this->execute(method: 'TRACE', headers: ['x-action' => 'requestHeadOrTrace']);
+        if($this->runtime === 'php') {
+            // Bug in PHP Swoole
+            self::assertEquals(400, $response['code']);
+        } else {
+            self::assertEquals(200, $response['code']);
+        }
+
+        self::assertStringContainsString('TRACE', $response['headers']['x-open-runtimes-logs']);
```
Comment (Meldiron) [inline tests/Base.php:246]: Based on: https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/TRACE

Seems like TRACE should be allowed to have body. What's the reason to use logs here?

File: tests.sh [hunk 6]
```diff
@@ -1,13 +1,40 @@
...
+
+if ! command -v docker &> /dev/null; then
+    echo "Docker is not installed or running!"
+    exit 1
+fi
+
+if ! command -v timeout &> /dev/null; then
+    echo "timeout is not available."
+    exit 1
+fi
```
[the remark this answers is not in the document]
Comment (ItzNotABug) [reply tests.sh:20]: I think the `timout` isn't directly available on a Mac system. Needs to be added as a separate lib from `brew`. Or did I miss something?

File: tests/Base.php (hunk 5 above)
Comment (ItzNotABug) [reply tests/Base.php:246]: If I remember correctly, we discussed using the logs for both `Head` & `TRACE` as one of the runtimes did not return a body for `TRACE` as well. I will find the exact convo. and ping it again to you.

File: tests/Base.php (hunk 4 above)
Comment (ItzNotABug) [reply tests/Base.php:17]: Makes sense. Will check.

File: tests/Base.php [hunk 7]
```diff
@@ -207,14 +212,40 @@ public function testRequestMethod(): void
...
         self::assertEquals('PATCH', $response['body']);
     }
 
+    public function testHeadAndTraceRequestMethods(): void
+    {
+        $response = $this->execute(method: 'HEAD', headers: ['x-action' => 'requestHeadOrTrace']);
+        self::assertEquals(200, $response['code']);
+
+        // head, trace dont work well with cpp [drogon].
+        if($this->runtime === 'cpp') return;
```
Comment (Meldiron) [inline tests/Base.php:234]: Not sure if I like this, I would rather have 2 larget if/else, so I know what exactly is expected for CPP. Here it seems to be expected to give 200, but not proper logs.

**Review by Meldiron — COMMENTED** [review]
Last review, good job! Let's address remaining comments and Ill mark as approved

File: tests/Base.php (hunk 7 above)
Comment (ItzNotABug) [reply tests/Base.php:234]: Only the response-code works for CPP, so for that runtime, I returned instead of moving ahead with other tests as they'd eventually be skipped due to an `if/else` check. But I'll add anyway. Better if we have descriptive logs/comments.

**Comment by ItzNotABug** [comment]
> Last review, good job! Let's address remaining comments and Ill mark as approved

Will check the review comments and the failing php tests too. Thanks.

File: tests.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (Meldiron) [reply tests.sh:20]: In one of the main PRs for next release, we removed `timeout` logic from this test file: https://github.com/open-runtimes/open-runtimes/blob/feat-log-files/tests.sh

It didn't work well for some scenarios, it was flaky. Doing simple sleep worked better.

File: tests/Base.php (hunk 5 above)
Comment (Meldiron) [reply tests/Base.php:246]: Thank you so much for taking time to ensure quality of this repository 🙏
Let's switch back to body response, and have `if` to skip the test for the runtime that doesn't support it.

Later we can fix those separately, but using proper body test ensures quality for the rest of runtimes.

File: tests/Base.php (hunk 7 above)
Comment (Meldiron) [reply tests/Base.php:234]: Understandable, I would prefer this structrure of the test then:

```
if runtiem is cpp
	call()
	ensure 200()
else
	call()
	ensure 200()
	more tests...
```

Its a bit of code duplication but it's for edge-case that isn't supposed to be there, so it's good it stands out.

File: tests.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (ItzNotABug) [reply tests.sh:20]: I see, thanks. I'll move to using `sleep`.

File: tests/Base.php (hunk 5 above)
Comment (ItzNotABug) [reply tests/Base.php:246]: Sure thing.

File: tests/Base.php (hunk 7 above)
Comment (ItzNotABug) [reply tests/Base.php:234]: Agree. Will address this too.

File: tests.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (Meldiron) [reply tests.sh:20]: No need, for now we can just avoid adding this check for timeout command being available. Later when that other PR is merged (very soon), the sleep will be in use.

File: tests/Base.php (hunk 7 above)
Comment (ItzNotABug) [reply tests/Base.php]: Done.

File: tests/Base.php (hunk 5 above)
Comment (ItzNotABug) [reply tests/Base.php]: As discussed, most likely the web servers wouldn't really return anything. Hence the response body check is not added.

File: tests/Base.php (hunk 4 above)
Comment (ItzNotABug) [reply tests/Base.php]: Done.

File: tests.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (ItzNotABug) [reply tests.sh]: Done.

**Review by Meldiron — APPROVED** [review]

File: runtimes/kotlin-1.6/src/main/kotlin/io/openruntimes/kotlin/Server.kt [hunk 8]
```diff
@@ -28,6 +29,7 @@ suspend fun main() {
         .patch("/*") { runBlocking { execute(it) } }
         .options("/*") { runBlocking { execute(it) } }
         .head("/*") { runBlocking { execute(it) } }
+        .addHandler(HandlerType.TRACE, "/*") { runBlocking { execute(it) } }
```
Comment (christyjacob4) [inline runtimes/kotlin-1.6/src/main/kotlin/io/openruntimes/kotlin/Server.kt:32]: as per the docs, https://javalin.io/documentation#handlers you can use `trace()` directly.

File: runtimes/kotlin-1.8/src/main/kotlin/io/openruntimes/kotlin/Server.kt [hunk 9]
```diff
@@ -28,6 +29,7 @@ suspend fun main() {
         .patch("/*") { runBlocking { execute(it) } }
         .options("/*") { runBlocking { execute(it) } }
         .head("/*") { runBlocking { execute(it) } }
+        .addHandler(HandlerType.TRACE, "/*") { runBlocking { execute(it) } }
```
Comment (christyjacob4) [inline runtimes/kotlin-1.8/src/main/kotlin/io/openruntimes/kotlin/Server.kt:32]: as per the docs, https://javalin.io/documentation#handlers you can use `trace()` directly.

File: runtimes/kotlin-1.9/src/main/kotlin/io/openruntimes/kotlin/Server.kt [hunk 10]
```diff
@@ -28,6 +29,7 @@ suspend fun main() {
         .patch("/*") { runBlocking { execute(it) } }
         .options("/*") { runBlocking { execute(it) } }
         .head("/*") { runBlocking { execute(it) } }
+        .addHandler(HandlerType.TRACE, "/*") { runBlocking { execute(it) } }
```
Comment (christyjacob4) [inline runtimes/kotlin-1.9/src/main/kotlin/io/openruntimes/kotlin/Server.kt:32]: as per the docs, https://javalin.io/documentation#handlers you can use `trace()` directly.

File: runtimes/python-3.10/src/server.py [hunk 11]
```diff
@@ -69,7 +69,7 @@ def error(self, message):
         else:
             self.errors.append(str(message))
 
-HTTP_METHODS = ['GET', 'HEAD', 'POST', 'PUT', 'DELETE', 'CONNECT', 'OPTIONS', 'TRACE', 'PATCH']
+HTTP_METHODS = ['GET', 'HEAD', 'POST', 'PUT', 'DELETE', 'OPTIONS', 'TRACE', 'PATCH']
```
Comment (christyjacob4) [inline runtimes/python-3.10/src/server.py:72]: why are we removing this ?

File: runtimes/ruby-3.3/src/server.rb [hunk 12]
```diff
@@ -310,6 +319,10 @@ def handle(request, response)
   handle(request, response)
 end
 
+trace '*' do
+  handle(request, response)
+end
+
```
Comment (christyjacob4) [inline runtimes/ruby-3.3/src/server.rb:325]: what about head requests for ruby? I dont see a handler for it?

**Review by christyjacob4 — CHANGES_REQUESTED** [review]
Awesome Work!! 👏  Just a couple of questions

File: runtimes/kotlin-1.6/src/main/kotlin/io/openruntimes/kotlin/Server.kt (hunk 8 above)
Comment (ItzNotABug) [reply runtimes/kotlin-1.6/src/main/kotlin/io/openruntimes/kotlin/Server.kt:32]: Docs mention supporting `TRACE`, however the routing api doesn't have a dedicated `trace()` method. I actually tried it earlier, the tests failed as there was no method. Ref - https://github.com/open-runtimes/open-runtimes/pull/242/commits/5f6085239ea45cbca691eca2fd6bf56cd80658d5

File: runtimes/python-3.10/src/server.py (hunk 11 above)
Comment (ItzNotABug) [reply runtimes/python-3.10/src/server.py:72]: If I recall correctly, it was to keep consistent methods across runtimes. Not all of them support `CONNECT`.

File: runtimes/ruby-3.3/src/server.rb (hunk 12 above)
Comment (ItzNotABug) [reply runtimes/ruby-3.3/src/server.rb:325]: The `get` handler auto-registers one for `head` too, [here's the reference](https://github.com/sinatra/sinatra/blob/main/lib/sinatra/base.rb#L1524C7-L1532C10).

Its been a while since I tested ruby, but the `head` handler's request are probably overriden with the `get`.
